### DISCUSSION TITLE
Simulating elasticity, expansion and stress on a solid whilst using Navier Stokes

### DISCUSSION THREAD

**Discussion by oh304** [idaholab/moose#27732] [Q&A Modules: Solid mechanics]
Hey! I've been using Navier Stokes to simulate fluid flow in a reactor, which has so far been a success - thanks! I am now interested to see  the stresses and other thermal effects experienced by the solid. 

I have two INSFV variables for temperature - one representing the temperature of the fluid, and the other for the solid.

    [T]                                     
        type = INSFVEnergyVariable
        block = 'fuel_salt'
        initial_condition = ${fuel_initial_temperature}
    []
    [Ts]                                    
        type = INSFVEnergyVariable
        block = 'inconel_outer inconel_inner'
        initial_condition = ${inconel_pipe_initial_temperature}
    []

I have implemented the SolidMechanics physics as such:

[Physics]
    [SolidMechanics]
      [QuasiStatic]
        [./all]
          strain = SMALL
          incremental = true
          add_variables = true
          eigenstrain_names = eigenstrain
          generate_output = 'strain_xx strain_yy'
        [../]
      [../]
    [../]
[]

And the thermal effects in the materials section:

    [elasticity_inconel]
        type = ADComputeIsotropicElasticityTensor
        youngs_modulus = 1e4
        poissons_ratio = 0.3
        block = 'inconel_outer inconel_inner'
    []
    [expansion_inconel]
        type = ADComputeThermalExpansionEigenstrain 
        temperature = 'Ts'
        thermal_expansion_coeff = 1e-3
        stress_free_temperature = 30
        eigenstrain_name = thermal_expansion
        block = 'inconel_outer inconel_inner'
    []
    [stress_inconel]
        type = ADComputeFiniteStrainElasticStress
        block = 'inconel_outer inconel_inner'
    []

Is there anything I seem to be missing? The error I receive is 'Variable 'temperature' is of a different C++ type than you tried to fetch it as'. My apologies if I have missed something obvious!

Best wishes,

Oscar

**Comment by GiudGiud** [comment]
Hello

The solid mechanics physics is expecting a finite element variable, and you are defining it as a finite volume variable

You can either set up a multiapp for the solid thermo-mechanics or use different variables with separate domains of definition for the fluid and solid temperatures.
The temperature in the solid should be a finite element variable, and in the fluid a finite volume variable (what you have) if you use a finite volume discretization of the flow equations.

Solving with two variables for temperature, one finite volume and one finite element should be possible. The connection should be done with boundary conditions or interface kernels 

I think there are quite a few examples of thermo-mechanics simulations with flow in a coolant channel (albeit using NekRS for flow instead of INSFV) in Cardinal (open-source application: https://github.com/neams-th-coe/cardinal/). That could be an inspiration

Guillaume
